### DISCUSSION TITLE
Question about Version control with KiBot : updating the KiCAD version/revision from KiBOT or script

### DISCUSSION THREAD

**Discussion by Strooom** [INTI-CMNB/KiBot#679] [Q&A] [answered]
Hi, for my firmware development I use a standard [semver ](https://semver.org/) approach where the CI workflow automatically increments the version. (eg https://github.com/Strooom/MuMo-v2-Node-SW)

Now I would like to do a similar thing for my HW/KiCad designs where (depending on the commit message/title) the workflow increments the version **and then puts this new version into the design**. So this means I would like to **update** the version fields in the KiCAD project files from KiBOT. Would that be possible somehow, or do I need an external script manipulate the KiCAD files for that ?

Thank you.

**Comment by set-soft** [comment, marked as answer]
You can use the `set_text_variables` preflight.

https://kibot.readthedocs.io/en/master/configuration/preflights/set_text_variables.html

I usually define the variable in KiCad with a value like *WIP*, then you can distinguish easily something generated from KiCad manually from something generated by a workflow.

The deprecated preflights contains some ideas about what to use as command.

https://kibot.readthedocs.io/en/master/configuration/preflights/PCB_ReplaceOptions.html#pcb-replaceoptions

**Reply by Strooom** [reply]
Perfect. Thanks

**Reply by Strooom** [reply]
So my github action workflow contains some section that determines the version we are building/releasing. But I need to get those parameters into the KiBOT yaml file.
I could have the workflow generate a .yaml file, which is then imported into the main KiBOT yaml file.

But maybe there is a simpler solution ? Can they be passed in the KiBOT github action as 'additional_args' :?

**Reply by set-soft** [reply]
KiBot can run any command you specify, you can use git tags and then run the git command to retrieve the tag.
You can also put the value in a file and then just use `echo FILE` as command.
You can also use environment variables, using the *env* section of your workflow (consult the GitHub docs)

**Comment by Strooom** [comment]
Ok, I was able to get a bit further but am still stuck on some syntax problems:
- in the github action I prepare a version.yaml file containing some 'definitions'
```
definition:
  majorVersionDigit: 0
  minorVersionDigit: 3

```
- I then include this file in the main config file for KiBOT
```
import:
  - version.yaml
```
- in preflight, I use 
```
preflight:
  set_text_variables:
    - variable: "majorVersion"
      text: "@majorVersionDigit@" 
    - variable: "minorVersion"
      text: "@minorVersionDigit@"
```
and in KiCAD I can then use ${majorVersion} and ${minorVersion} to put the version on the silkscreen..

But the substitution in "@majorVersionDigit@" is not working, ie. the variable is getting the value "@majorVersionDigit@" instead of  "0".
I've tried without enclosing quotes @majorVersionDigit@ but this results in a yaml error as a key value cannot start with '@'.

Any hints on how to get this working ?

**Reply by set-soft** [reply]
This doesn't look well. 
1. The keyword is `definitions`, not `definition` and used for a very different stuff, read the docs.
2. I don't know from where comes the values you assign to `majorVersionDigit` by I strongly recommend to get them from the current tag in the repo using the `git` command (i.e. `git describe --tags --abbrev=0`). Or from a file in the repo. Avoid substitutions when you can. So use the `command` not `text`.

**Reply by Strooom** [reply]
Yes, those numbers are derived from the latest tag in the repo, but in order to determine what will be the **next** version, there is more logic involved, eg. what branch are we on, is this a major or minor new release.... 
Here is the part of the action to extract data
```
run: |
          echo "extract branch name from github_ref '${{ github.ref }}'"
          declare branchname=$(echo "${{ github.ref }}" | cut -d'/' -f 3-)
          echo "clean branch name = $branchname"
          echo "extract commit short hash : $(git rev-parse --short HEAD)"
          declare commithash=$(git rev-parse --short HEAD)
          echo "extract build timestamp"
          declare buildtimestamp=$(date +"%a %b %d %H:%M:%S %Y")
          echo "buildtimestamp = $buildtimestamp"
          declare -i buildEpoch=$(date "+%s")

          declare fulltag=$(git describe --tag $(git rev-parse --verify refs/remotes/origin/main))
          echo "fulltag = [$fulltag]"
          declare versiontag=$(echo $fulltag | cut -d'-' -f1)
          echo "extract SemVer numbers from version tag [$versiontag]"
          declare -i lastmajordigit=$(echo $versiontag | cut -c 2- | cut -d'.' -f1)
          echo "lastmajordigit = $lastmajordigit"
          declare -i lastminordigit=$(echo $versiontag | cut -c 2- | cut -d'.' -f2)
          echo "lastminordigit = $lastminordigit"
          declare lastversion="v$lastmajordigit.$lastminordigit"
          echo "output variables to GitHub Actions"
          echo "branchname=$branchname" >> $GITHUB_OUTPUT
          echo "lastmajordigit=$lastmajordigit" >> $GITHUB_OUTPUT
          echo "lastminordigit=$lastminordigit" >> $GITHUB_OUTPUT
          echo "commithash=$commithash" >> $GITHUB_OUTPUT
          echo "buildtimestamp=$buildtimestamp" >> $GITHUB_OUTPUT
          echo "buildEpoch=$buildEpoch" >> $GITHUB_OUTPUT
          echo "lastversion=$lastversion" >> $GITHUB_OUTPUT
```
Then, after extracting what is the properties of the last release, here is the code to derive the next release..
```
       run: |
          echo "Triggered from Branch : ${{ steps.versioninfo.outputs.branchname }}"
          echo "Commit hash           : ${{ steps.versioninfo.outputs.commithash }}"
          echo "Last version          : ${{ steps.versioninfo.outputs.lastversion }}"
          echo "  Major               : ${{ steps.versioninfo.outputs.lastmajordigit }}"
          echo "  Minor               : ${{ steps.versioninfo.outputs.lastminordigit }}"
          if [ "${{ steps.versioninfo.outputs.branchname }}" == "main" ]; then
            echo "Triggered from merge on main branch with commit title : ${{ github.event.head_commit.message }}"
            if [[ "${{ github.event.head_commit.message }}" == *"major"* ]]; then
              echo "Incrementing Major versionDigit"
              declare -i nextmajordigit=${{ steps.versioninfo.outputs.lastmajordigit }}+1
              declare -i nextminordigit=0
              declare buildversion="v$nextmajordigit.$nextminordigit"
            else
              echo "Incrementing Minor versionDigit"
              declare -i nextmajordigit=${{ steps.versioninfo.outputs.lastmajordigit }}
              declare -i nextminordigit=${{ steps.versioninfo.outputs.lastminordigit }}+1
              declare buildversion="v$nextmajordigit.$nextminordigit"
            fi
          else
            echo "Not on main branch -> development version"
            declare -i nextmajordigit=${{ steps.versioninfo.outputs.lastmajordigit }}
            declare -i nextminordigit=${{ steps.versioninfo.outputs.lastminordigit }}
            declare buildversion="v$nextmajordigit.$nextminordigit-${{ steps.versioninfo.outputs.commithash }}"
          fi
          echo "Building Version : $buildversion"
          echo "  Major          : $nextmajordigit"
          echo "  Minor          : $nextminordigit"
          echo "output variables to GitHub Actions"
          echo "nextmajordigit=$nextmajordigit" >> $GITHUB_OUTPUT
          echo "nextminordigit=$nextminordigit" >> $GITHUB_OUTPUT
          echo "buildversion=$buildversion" >> $GITHUB_OUTPUT
```
Then, to have a simple checkpoint, I write all this to a file, in this case version.yaml
So what is the best way to get the info from this file into KiCAD variables ?

**Reply by set-soft** [reply]
You can simply write each number to a separated file and use something like `cat FILE` as command for the `set_text_variables`, or even use environment variables, you can run **anything you want** to get the value for the variable.

**Reply by Strooom** [reply]
Ok, using environment variables I got it working : setting them in the github action workflow like
```
      - name: Save info in environment variables
        run: |
          echo "MAJORVERSIONDIGIT=${{ steps.selectversion.outputs.nextmajordigit }}" >> $GITHUB_ENV
          echo "MINORVERSIONDIGIT=${{ steps.selectversion.outputs.nextminordigit }}" >> $GITHUB_ENV
```
Then using them in KiBOT like this
```
preflight:
  set_text_variables:
    - variable: "majorVersion"
      command: "echo $MAJORVERSIONDIGIT" 
    - variable: "minorVersion"
      command: "echo $MINORVERSIONDIGIT" 
```
Thank you for your support !
